File: app/src/main/java/ardjomand/leonardo/arcmovies/data/MoviesRepository.java
package ardjomand.leonardo.arcmovies.data;

import java.util.List;

import ardjomand.leonardo.arcmovies.model.Genre;
import ardjomand.leonardo.arcmovies.model.MovieDetails;
import ardjomand.leonardo.arcmovies.model.UpcomingMovies;

/**
 * Created by unity on 25/02/18.
 */

public interface MoviesRepository {

    interface LoadUpcomingMoviesCallback {

        void onSuccess(UpcomingMovies upcomingMovies);

        void onFailure(String message);
    }

    void loadUpcomingMovies(LoadUpcomingMoviesCallback loadUpcomingMoviesCallback, int page);


    interface LoadMovieDetailsCallback {

        void onSuccess(MovieDetails movieDetails);

        void onFailure(String message);
    }

    void loadMovieDetails(LoadMovieDetailsCallback loadMovieDetailsCallback, int movieId);


    interface LoadGenresCallBack {

        void onSuccess(List<Genre> genres);

        void onFailure(String message);
    }

    void loadGenres(LoadGenresCallBack loadGenresCallBack);
}

File: app/src/main/java/ardjomand/leonardo/arcmovies/data/MoviesRepository.kt
package ardjomand.leonardo.arcmovies.data

import ardjomand.leonardo.arcmovies.model.Genre
import ardjomand.leonardo.arcmovies.model.MovieDetails
import ardjomand.leonardo.arcmovies.model.UpcomingMovies

/**
 * Created by unity on 25/02/18.
 */

interface MoviesRepository {

    interface LoadUpcomingMoviesCallback {

        fun onSuccess(upcomingMovies: UpcomingMovies)

        fun onFailure(message: String)
    }

    fun loadUpcomingMovies(loadUpcomingMoviesCallback: LoadUpcomingMoviesCallback, page: Int)


    interface LoadMovieDetailsCallback {

        fun onSuccess(movieDetails: MovieDetails)

        fun onFailure(message: String)
    }

    fun loadMovieDetails(loadMovieDetailsCallback: LoadMovieDetailsCallback, movieId: Int)


    interface LoadGenresCallBack {

        fun onSuccess(genres: List<Genre>)

        fun onFailure(message: String)
    }

    fun loadGenres(loadGenresCallBack: LoadGenresCallBack)
}
